src/Proper/HasPermutationGenerator.hs

```diff
@@ -0,0 +1,301 @@
+module Proper.HasPermutationGenerator (
+  HasPermutationGenerator(..),
+  PermutationEdge(..),
+  ) where
+import Debug.Trace
+import Proper.HasLogicalModel
+import Proper.LogicalModel
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Hedgehog (Gen,PropertyT,MonadTest,Group(..),forAll,failure,footnote,property)
+import qualified Hedgehog.Gen as Gen
+import Data.Map (Map)
+import qualified Data.Map as Map
+import SAT.MiniSat (Formula(..))
+import Data.Proxy (Proxy(..))
+import Data.Graph (Graph)
+import Data.Graph (buildG,scc,path)
+import Text.Show.Pretty (ppDoc)
+import Text.PrettyPrint (
+  Style (lineLength),
+  hang,
+  renderStyle,
+  style,
+  ($+$),
+ )
+import Control.Monad (join)
+import Data.String (fromString)
+
+data PermutationEdge m p =
+  PermutationEdge {
+    name :: String
+  , match :: Formula p
+  , contract :: Set p -> Set p
+  , permuteGen :: m -> Gen m
+  }
+
+instance Eq (PermutationEdge m p) where
+  (==) a b = name a == name b
+
+instance Show (PermutationEdge m p) where
+  show = name
+
+class (HasLogicalModel m p, Show m) => HasPermutationGenerator m p where
+  generators :: [PermutationEdge m p]
+
+  permutationGeneratorSelfTest :: (PermutationEdge m p -> Bool) -> Gen m -> [Group]
+  permutationGeneratorSelfTest pefilter bgen =
+    let pedges = findPermutationEdges (Proxy :: Proxy m) (Proxy :: Proxy p)
+        (_,ns) = numberNodes (Proxy :: Proxy m) (Proxy :: Proxy p)
+        mGen = buildGen bgen
+        graph = buildGraph pedges
+        isco = isStronglyConnected graph
+     in if isco
+           then case findDupEdgeNames of
+                  [] -> testEdge ns pedges mGen <$> filter pefilter generators
+                  dups -> [Group "HasPermutationGenerator edge names must be unique." $
+                           [(fromString $ dup <> " not unique", property $ failure)
+                           | dup <- dups]
+                          ]
+           else [Group "HasPermutationGenerator Graph Not Strongly Connected" $
+                          [(fromString "Not strongly connected", abortNotSCC ns graph)]
+                        ]
```

Comment:
The `PermutationGenerator` self-test suite checks that the edge names are unique and the graph is strongly connected. If the graph is not strongly connected we get a hint as to which edges need to be defined.